staging/src/k8s.io/component-base/featuregate/testing/feature_gate.go

```diff
@@ -18,18 +18,46 @@ package testing
 
 import (
 	"fmt"
-	"testing"
-
 	"k8s.io/component-base/featuregate"
+	"sync"
+)
+
+var (
+	overwrittenFeatureFlags map[featuregate.Feature]struct{}
+	lock                    sync.Mutex
 )
 
-// SetFeatureGateDuringTest sets the specified gate to the specified value, and returns a function that restores the original value.
+func init() {
+	overwrittenFeatureFlags = map[featuregate.Feature]struct{}{}
+}
+
+// SetFeatureGateDuringTest sets the specified gate to the specified value.
 // Failures to set or restore cause the test to fail.
 //
 // Example use:
 //
-// defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, true)()
-func SetFeatureGateDuringTest(tb testing.TB, gate featuregate.FeatureGate, f featuregate.Feature, value bool) func() {
+// featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, true)
+func SetFeatureGateDuringTest(tb testingT, gate featuregate.FeatureGate, f featuregate.Feature, value bool) {
+	lock.Lock()
+	defer lock.Unlock()
+	if _, ok := overwrittenFeatureFlags[f]; ok {
+		tb.Fatal("Detected parallel setting of a feature gate")
+	}
```

Comment:
adding detection seems good here

I'm -1 on churning a signature change for this through the codebase on code freeze day

I'm on the fence about requiring the cleanup to be tied to t.Cleanup ... do we have uses inside functions whose lifecycles are shorter than the test lifecycle?

Reply:
> I'm on the fence about requiring the cleanup to be tied to t.Cleanup ... do we have uses inside functions whose lifecycles are shorter than the test lifecycle?

Subtests that call t.Parallel have livecycle shorter than the test. And unfortunatelly we have those https://github.com/kubernetes/kubernetes/pull/123676/files#r1511729221

Reply:
Testing cacher correctness depends on feature flag https://github.com/kubernetes/kubernetes/blob/7a20def5ba9f8e399f21467a194e85f21cbd6a47/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_test.go#L175-L180
and parallelization is important https://github.com/kubernetes/kubernetes/blob/7a20def5ba9f8e399f21467a194e85f21cbd6a47/staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go#L203-L210

Reply:
seems like tests running Parallel trying to use this have to rework themselves anyway ... let's limit this PR to detection of conflicts for now, can revisit after code freeze / before test freeze

Reply:
> I'm -1 on churning a signature change for this through the codebase on code freeze day

Makes sense, still would be good to know if any other feature flags tests are broken.

Reply:
Proposed to skip changing the function signature. We can just keep the defer by returning an empty function. It should maintain the same usage semantic, override flag within the test body.